### PR TITLE
pkg/trace/traceutil: refactor ComputeTopLevel for readability

### PR DESCRIPTION
**Pull request by martpier**
### What does this PR do?

Clarify documentation and comment of `ComputeTopLevel`.

### Motivation

When initially reading this function:
 - I did not understand immediately that bullet points were to be interpreted as "OR" rather than "AND"
 - I struggled understanding the code

### Reviewer's Checklist

- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] Changed code has automated tests for its functionality.
- [x] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [x] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [x] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [x] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [x] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: pkg/trace/traceutil/trace.go [hunk 1]
```diff
@@ -109,24 +109,35 @@ func ChildrenMap(t pb.Trace) map[uint64][]*pb.Span {
...
 
-	// iterate on each span and mark them as top-level if relevant
-	for _, span := range t {
-		if span.ParentID != 0 {
-			if parentIdx, ok := spanIDToIdx[span.ParentID]; ok && t[parentIdx].Service == span.Service {
-				continue
-			}
+	for _, span := range trace {
+		isRootSpan := span.ParentID == 0
+
```
Comment (dumontg) [inline pkg/trace/traceutil/trace.go:124]: nit on code style: I like spaces to separate logical blocks. But to me L123 and L125 are very much connected. Same for L131 and L138, they are just making it harder to read IMO.

Comment (martpier) [reply pkg/trace/traceutil/trace.go]: I agree, fixed ;)

File: pkg/trace/traceutil/trace.go [hunk 2]
```diff
@@ -109,24 +109,32 @@ func ChildrenMap(t pb.Trace) map[uint64][]*pb.Span {
...
+// - OR its parent belongs to another service (in that case it's a "local root"
 //   being the highest ancestor of other spans belonging to this service and
 //   attached to it).
-func ComputeTopLevel(t pb.Trace) {
-	// build a lookup map
-	spanIDToIdx := make(map[uint64]int, len(t))
-	for i, span := range t {
-		spanIDToIdx[span.SpanID] = i
+func ComputeTopLevel(trace pb.Trace) {
+	spanIdToIndex := make(map[uint64]int, len(trace))
```
Comment (gbbr) [inline pkg/trace/traceutil/trace.go:117]: This is fine, but please keep abbreviations all caps
```suggestion
	spanIDToIndex := make(map[uint64]int, len(trace))
```

File: pkg/trace/traceutil/trace.go [hunk 3]
```diff
@@ -109,24 +109,32 @@ func ChildrenMap(t pb.Trace) map[uint64][]*pb.Span {
...
 	}
 
-	// iterate on each span and mark them as top-level if relevant
-	for _, span := range t {
-		if span.ParentID != 0 {
-			if parentIdx, ok := spanIDToIdx[span.ParentID]; ok && t[parentIdx].Service == span.Service {
-				continue
-			}
+	for _, span := range trace {
+		isRootSpan := span.ParentID == 0
```
Comment (gbbr) [inline pkg/trace/traceutil/trace.go:123]: I personally don't think code should be documented through the means of variable names. If you want to do that, just do:
```go
if span.ParentID == 0 {
    // root spans are top level
    ...
}
```
And not name variables to document things.

File: pkg/trace/traceutil/trace.go [hunk 4]
```diff
@@ -109,24 +109,32 @@ func ChildrenMap(t pb.Trace) map[uint64][]*pb.Span {
...
+		if isRootSpan {
+			SetTopLevel(span, true)
+			continue
+		}
+
+		parentIndex, isParentKnown := spanIdToIndex[span.ParentID]
+		if !isParentKnown {
```
Comment (gbbr) [inline pkg/trace/traceutil/trace.go:130]: Same here. The convention in Go for checking if a map item exists is naming it `ok`
```suggestion
		parentIndex, ok := spanIDToIndex[span.ParentID]
		if !ok {
		   // has no parent in chunk
		   ...
		}
```

File: pkg/trace/traceutil/trace.go [hunk 5]
```diff
@@ -109,24 +109,32 @@ func ChildrenMap(t pb.Trace) map[uint64][]*pb.Span {
...
+			SetTopLevel(span, true)
+			continue
+		}
+
+		isParentInSameService := trace[parentIndex].Service == span.Service
+		if !isParentInSameService {
```
Comment (gbbr) [inline pkg/trace/traceutil/trace.go:136]: Same here:
```suggestion
		if trace[parentIndex].Service != span.Service {
		    // parent is not the same service
```

File: pkg/trace/traceutil/trace.go [hunk 6]
```diff
@@ -109,24 +109,32 @@ func ChildrenMap(t pb.Trace) map[uint64][]*pb.Span {
...
-			if parentIdx, ok := spanIDToIdx[span.ParentID]; ok && t[parentIdx].Service == span.Service {
-				continue
-			}
+	for _, span := range trace {
+		isRootSpan := span.ParentID == 0
+		if isRootSpan {
+			SetTopLevel(span, true)
+			continue
+		}
+
```
Comment (gbbr) [inline pkg/trace/traceutil/trace.go:128]: We use the same style guidelines in `pkg/trace` as we do in `dd-trace-go`: https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines

Can you please remove the empty lines added? (check nr. 3 in "[General Guidelines](https://github.com/DataDog/dd-trace-go/wiki/Style-guidelines#general-guidelines)")

**Review by gbbr — CHANGES_REQUESTED** [review]
This is a nice improvement, it helps. But please don't use variable names to document code. Document code through comments instead.

Lastly, please name the PR according to convention, by prefixing it with the package name that is being changed, followed by a colon and by a lower-case sentence. For example:

```go
pkg/trace/traceutil: refactor ComputeTopLevel for readability
```

Generally, I don't think it's ok to change such a short function which is so critical to the correct functioning of the product and risk regressions.

File: pkg/trace/traceutil/trace.go [hunk 7]
```diff
@@ -109,24 +109,32 @@ func ChildrenMap(t pb.Trace) map[uint64][]*pb.Span {
...
+		parentIndex, isParentKnown := spanIdToIndex[span.ParentID]
+		if !isParentKnown {
+			SetTopLevel(span, true)
+			continue
+		}
+
+		isParentInSameService := trace[parentIndex].Service == span.Service
+		if !isParentInSameService {
+			SetTopLevel(span, true)
```
Comment (gbbr) [inline pkg/trace/traceutil/trace.go:137]: Technically you don't need the `continue` call anywhere, but it's fine that you've added it. It is safer. For the same reason, should we also add it here?

File: pkg/trace/traceutil/trace.go (hunk 2 above)
Comment (martpier) [reply pkg/trace/traceutil/trace.go]: fixed :)

File: pkg/trace/traceutil/trace.go (hunk 6 above)
Comment (martpier) [reply pkg/trace/traceutil/trace.go]: thanks for the detailed reference, fixed :)

File: pkg/trace/traceutil/trace.go (hunk 7 above)
Comment (martpier) [reply pkg/trace/traceutil/trace.go:135]: fixed

File: pkg/trace/traceutil/trace.go (hunk 3 above)
Comment (martpier) [reply pkg/trace/traceutil/trace.go]: fixed

**Review by gbbr — APPROVED** [review]
Thanks! This is a good improvement.

**Pull request merged**